lua/starfall/preprocessor.lua

```diff
@@ -123,23 +121,148 @@ function SF.Preprocessor.ParseDirectives ( filename, source, directives, data, i
 				endingLevel = comment.level
 			elseif comment.type == "line" then
 				local directive, args = string.match( line, "--@([^ ]+)%s*(.*)$" )
-				local func = directives[ directive ] or SF.Preprocessor.directives[ directive ]
-				if func then
-					func( args, filename, data, instance )
+				if directive then
+					directives[ directive ] = args
 				end
 			end
 		end
-		
+
 		if ending == "newline" then ending = nil end
 	end
+	return directives
+end
+
+--- Parses a source file for directives
+-- @param filename The file name of the source code.
+-- @param source The source code to parse.
+-- @param preprocs The preprocessors to search for, can be a string for one, a table for multiple, and nil for all
+-- @return Table of preprocs and their args
+function SF.Preprocessor.GetDirectives ( filename, source, preprocs )
+	local parsedDirectives = parseDirectives( filename, source )
+
+	local preprocs_type = type( preprocs )
+	if preprocs_type == "string" then --1 specified preproc
+		return parsedDirectives[ preprocs ]
+	elseif preprocs_type == "table" then --Table of specified preprocs
+		local ret = {}
+		for _, preproc in pairs( preprocs ) do
+			ret[ preproc ] = parsedDirectives[ preproc ]
+		end
+		return ret
+	elseif preprocs == nil then --All preprocs in the file
+		return parsedDirectives
+	end
 end
 
-local function directive_include (args, filename, data)
-	if not data.includes then data.includes = {} end
-	if not data.includes[ filename ] then data.includes[ filename ] = {} end
-	
-	local incl = data.includes[ filename ]
-	incl[ #incl + 1 ] = args
+--- Parses a source file for directives
+-- @param filename The file name of the source code.
+-- @param source The source code to parse.
+-- @param directives A table of additional directives to use.
+-- @param data The data table passed to the directives.
+-- @param instance The instance
+-- @param preprocs The preprocessors to search for, can be a string for one, a table for multiple, and nil for all
+-- @return Table of preprocs and their args
+function SF.Preprocessor.ParseDirectives ( filename, source, directives, data, instance, preprocs )
+	local parsedDirectives = SF.Preprocessor.GetDirectives( filename, source )
+
+	local ok, err = true, {}
+
+	local preprocs_type = type( preprocs )
+	if preprocs_type == "string" then --1 specified preproc
+		local func = directives[ preprocs ] or SF.Preprocessor.directives[ preprocs ]
```

Comment:
Lots of code blocks like this switching on type. As suggested, keep it as nil or a table. Try and optimise this to a single block, make use of ternary for nil checks etc.